File: app/src/main/java/edu/tjrac/swant/bluetooth/view/ScanInfoHistoryFragment.java
package edu.tjrac.swant.bluetooth.view;

import android.annotation.SuppressLint;
import android.bluetooth.le.ScanResult;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.SpannableStringBuilder;
import android.text.style.ForegroundColorSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.yckj.baselib.common.views.BaseChartView;
import com.yckj.baselib.util.L;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import edu.tjrac.swant.unicorn.R;
import edu.tjrac.swant.bluetooth.BlueToothHelper;
import edu.tjrac.swant.bluetooth.bean.ScanInfo;
import edu.tjrac.swant.bluetooth.adapter.ScanInfoHistoryAdapter;

import static android.text.Spanned.SPAN_INCLUSIVE_INCLUSIVE;
import static edu.tjrac.swant.bluetooth.BlueToothHelper.keySet;

/**
 * Created by wpc on 2018/2/2 0002.
 */

@SuppressLint("ValidFragment")
public class ScanInfoHistoryFragment extends Fragment {

    @BindView(R.id.rssi_view) BaseChartView mRssiView;
    @BindView(R.id.recycler) RecyclerView mRecycler;
    ArrayList<ScanInfo> infos;
    ArrayList<BaseChartView.ChartValue> chartValues;
    ScanInfoHistoryAdapter adapter;
    Unbinder unbinder;
    private ScanResult result;


    ScanInfoHistoryFragment(ScanResult result, ArrayList<ScanInfo> infos) {
        this.infos = infos;
        this.result = result;
    }


    @SuppressLint("ResourceAsColor")
    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.scan_info_history, container, false);
        unbinder = ButterKnife.bind(this, v);
//        mRssiView.setTitles(32,"TIME(s)","RSSI(dBm)",getActivity().getResources().getColor(R.color.blue));
        return v;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {

        chartValues = getChartValues(infos);

        mRssiView.setStyle(new BaseChartView.StyleBuilder()
                .line(true)
                .dashPath(true)
                .setXInfo(10f, 0f, 10, "TIME(s)", getActivity().getResources().getColor(R.color.blue))
                .setYInfo(-100, 0, 5, "RSSI(dBm)", getActivity().getResources().getColor(R.color.blue))
                .addLine(getActivity().getResources().getColor(R.color.blue),
                        chartValues)
        );
        L.i("chartValue", chartValues.size() + "");
        mRecycler.setLayoutManager(new LinearLayoutManager(getActivity()));
        adapter = new ScanInfoHistoryAdapter(infos);

        adapter.bindToRecyclerView(mRecycler);
        View headview = LayoutInflater.from(getActivity()).inflate(R.layout.head_scan_infos, null);
        initHeadView(headview);
        adapter.addHeaderView(headview);
        mRecycler.setAdapter(adapter);
        super.onViewCreated(view, savedInstanceState);
    }


    String[] values;

    private void initHeadView(View headview) {
        values = new String[4];
        values[0] = BlueToothHelper.getType(result.getDevice().getType());
        boolean[] flags = BlueToothHelper.getFlags(result.getScanRecord().getAdvertiseFlags());

        values[1] = BlueToothHelper.getFlags(flags);
        values[2] = BlueToothHelper.getStringUUIDs(result.getScanRecord().getServiceUuids());
        values[3] = result.getDevice().getName();
        LinearLayout ll = (LinearLayout) headview.findViewById(R.id.infos);
        for (int i = 0; i < keySet.length; i++) {
            SpannableStringBuilder span = new SpannableStringBuilder(keySet[i] + values[i]);
            span.setSpan(new ForegroundColorSpan(getActivity().getResources().getColor(R.color.blue)),
                    0, keySet[i].length(), SPAN_INCLUSIVE_INCLUSIVE);
            TextView textView = new TextView(getActivity());
            textView.setText(span);
            ll.addView(textView);
        }
        ((TextView) headview.findViewById(R.id.div_time)).setText(infos.get(0).getAdvTime() + "ms");
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    private ArrayList<BaseChartView.ChartValue> getChartValues(ArrayList<ScanInfo> infos) {
        Long base = infos.get(0).getTagTime();
        ArrayList<BaseChartView.ChartValue> values = new ArrayList<>();
        for (ScanInfo info : infos) {
            values.add(new BaseChartView.ChartValue((base - info.getTagTime()) / 1000, info.getDbm()));
        }
        return values;
    }
}

File: app/src/main/java/edu/tjrac/swant/bluetooth/view/ScanInfoHistoryFragment.kt
package edu.tjrac.swant.bluetooth.view

import android.annotation.SuppressLint
import android.bluetooth.le.ScanResult
import android.os.Build
import android.os.Bundle
import android.support.annotation.RequiresApi
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.text.SpannableStringBuilder
import android.text.style.ForegroundColorSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView

import com.yckj.baselib.common.views.BaseChartView
import com.yckj.baselib.util.L

import java.util.ArrayList

import edu.tjrac.swant.unicorn.R
import edu.tjrac.swant.bluetooth.BlueToothHelper
import edu.tjrac.swant.bluetooth.bean.ScanInfo
import edu.tjrac.swant.bluetooth.adapter.ScanInfoHistoryAdapter

import android.text.Spanned.SPAN_INCLUSIVE_INCLUSIVE
import edu.tjrac.swant.bluetooth.BlueToothHelper.keySet
import kotlinx.android.synthetic.main.scan_info_history.*

/**
 * Created by wpc on 2018/2/2 0002.
 */

@SuppressLint("ValidFragment")
class ScanInfoHistoryFragment internal constructor(private val result: ScanResult, internal var infos: ArrayList<ScanInfo>) : Fragment() {
    internal var chartValues: ArrayList<BaseChartView.ChartValue>?=null
    internal var adapter: ScanInfoHistoryAdapter?=null


    @SuppressLint("ResourceAsColor")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.scan_info_history, container, false)
        //        mRssiView.setTitles(32,"TIME(s)","RSSI(dBm)",getActivity().getResources().getColor(R.color.blue));
        return v
    }

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        chartValues = getChartValues(infos)

        rssi_view.setStyle(BaseChartView.StyleBuilder()
                .line(true)
                .dashPath(true)
                .setXInfo(10f, 0f, 10, "TIME(s)", activity!!.resources.getColor(R.color.blue))
                .setYInfo(-100f, 0f, 5, "RSSI(dBm)", activity!!.resources.getColor(R.color.blue))
                .addLine(activity!!.resources.getColor(R.color.blue),
                        chartValues)
        )
        L.i("chartValue", chartValues!!.size.toString() + "")
        recycler.setLayoutManager(LinearLayoutManager(activity))
        adapter = ScanInfoHistoryAdapter(infos)

        adapter!!.bindToRecyclerView(recycler)
        val headview = LayoutInflater.from(activity).inflate(R.layout.head_scan_infos, null)
        initHeadView(headview)
        adapter!!.addHeaderView(headview)
        recycler.setAdapter(adapter)
        super.onViewCreated(view, savedInstanceState)
    }
    internal var values: Array<String?>?=null

    @RequiresApi(Build.VERSION_CODES.LOLLIPOP)
    private fun initHeadView(headview: View) {
        values = arrayOfNulls<String>(4)
        values!![0] = BlueToothHelper.getType(result.device!!.type)
        val flags = BlueToothHelper.getFlags(result.scanRecord!!.advertiseFlags)

        values!![1] = BlueToothHelper.getFlags(flags)
        values!![2] = BlueToothHelper.getStringUUIDs(result.scanRecord!!.serviceUuids)
        values!![3] = result.device.name
        val ll = headview.findViewById<View>(R.id.infos) as LinearLayout
        for (i in keySet.indices) {
            val span = SpannableStringBuilder(keySet[i] + values!![i])
            span.setSpan(ForegroundColorSpan(activity!!.resources.getColor(R.color.blue)),
                    0, keySet[i].length, SPAN_INCLUSIVE_INCLUSIVE)
            val textView = TextView(activity)
            textView.text = span
            ll.addView(textView)
        }
        (headview.findViewById<View>(R.id.div_time) as TextView).text = infos[0].advTime.toString() + "ms"
    }

    override fun onDestroyView() {
        super.onDestroyView()
    }

    private fun getChartValues(infos: ArrayList<ScanInfo>): ArrayList<BaseChartView.ChartValue> {
        val base = infos[0].tagTime
        val values = ArrayList<BaseChartView.ChartValue>()
        for (info in infos) {
            values.add(BaseChartView.ChartValue(((base - info.tagTime) / 1000).toFloat(), info.dbm.toFloat()))
        }
        return values
    }
}
